pkg/controllers/leaderworkerset_controller.go

```diff
@@ -537,51 +543,101 @@ func (r *LeaderWorkerSetReconciler) iterateReplicas(ctx context.Context, lws *le
-		nominatedName := fmt.Sprintf("%s-%d", lws.Name, index)
+
+	for idx := int32(0); idx < stsReplicas; idx++ {
+		nominatedName := fmt.Sprintf("%s-%d", lws.Name, idx)
 		// It can happen that the leader pod or the worker statefulset hasn't created yet
 		// or under rebuilding, which also indicates not ready.
-		if nominatedName != sortedPods[index].Name || (!noWorkerSts && nominatedName != sortedSts[index].Name) {
-			return false
+		if nominatedName != sortedPods[idx].Name || (!noWorkerSts && nominatedName != sortedSts[idx].Name) {
+			states[idx] = replicaState{
+				ready:   false,
+				updated: false,
+			}
 		}
 
-		podTemplateHash := revisionutils.GetRevisionKey(&sortedPods[index])
-		if !(podTemplateHash == revisionKey && podutils.PodRunningAndReady(sortedPods[index])) {
-			return false
-		}
+		leaderUpdated := revisionutils.GetRevisionKey(&sortedPods[idx]) == revisionKey
+		leaderReady := podutils.PodRunningAndReady(sortedPods[idx])
 
 		if noWorkerSts {
-			return true
+			states[idx] = replicaState{
+				ready:   leaderReady,
+				updated: leaderUpdated,
+			}
+			continue
+		}
+
+		stsExists := (int(idx) < len(sortedSts) && nominatedName == sortedSts[idx].Name)
+		if !stsExists {
+			continue
 		}
 
-		stsTemplateHash := revisionutils.GetRevisionKey(&sortedSts[index])
-		return stsTemplateHash == revisionKey && statefulsetutils.StatefulsetReady(sortedSts[index])
+		workersUpdated := revisionutils.GetRevisionKey(&sortedSts[idx]) == revisionKey
+		workersReady := statefulsetutils.StatefulsetReady(sortedSts[idx])
+
+		states[idx] = replicaState{
+			ready:   leaderReady && workersReady,
+			updated: leaderUpdated && workersUpdated,
+		}
 	}
 
-	var skip bool
-	var continuousReadyReplicas, lwsUnreadyReplicas int32
+	return states, nil
+}
 
-	for index := stsReplicas - 1; index >= 0; index-- {
-		replicaReady := processReplica(index)
-		skip = skip || !replicaReady
+func calculateLWSUnreadyReplicas(states []replicaState, lwsReplicas int32) int32 {
+	var unreadyCount int32
+	for idx := int32(0); idx < min(int32(len(states)), lwsReplicas); idx++ {
+		if !states[idx].ready || !states[idx].updated {
+			unreadyCount++
+		}
+	}
+	return unreadyCount
+}
 
-		if replicaReady && !skip {
-			continuousReadyReplicas++
+func calculateContinuousReadyReplicas(states []replicaState) int32 {
+	// Count ready replicas at tail (from last index down)
+	var continuousReadyCount int32
+	for idx := len(states) - 1; idx >= 0; idx-- {
+		if !states[idx].ready || !states[idx].updated {
+			break
 		}
-		if !replicaReady && index < *lws.Spec.Replicas {
-			lwsUnreadyReplicas++
+		continuousReadyCount++
+	}
+	return continuousReadyCount
+}
+
+// Checks state of not updated replicas and adjusts the rolling update parittion to prevent
+// availability drops below "maxUnavailable".
+func partitionAccountingForUnavailable(states []replicaState, desiredPartition int32) int32 {
+	// Desired partition already accounts for readiness of replicas above partition. Code below
```

Comment:
Can we calculate just in one place? Like refactoring the original part, would that make the code more readable?

Reply:
Do you mean trying to extend the `iterateReplicas` and adding more return values to the result as @Edwinhr716 suggested?